### ISSUE TITLE
Midnight Resistance - guru while loading

### ISSUE THREAD

**Issue by bnice** [dirkwhoffmann/vAmiga#434]
Hi,

Midnight Resistance is guru meditation 00000003.00C01570 while loading. It does this both Angels and CLS cracked versions. Note this concludes testing all the amiga apps I have :-)

Image:
http://archive.org/download/Commodore_Amiga_TOSEC_2012_04_10/Commodore_Amiga_TOSEC_2012_04_10.zip/Commodore%20Amiga%20%5BTOSEC%20v2012-04-10%5D%2FGames%2F%5BADF%5D%2FCommodore%20Amiga%20-%20Games%20-%20%5BADF%5D%20%28TOSEC%202012-02-27%29%2FMidnight%20Resistance%20%281990%29%28Ocean%29%5Bcr%20Angels%5D.zip

**Comment by bnice** [comment]
Note both versions work on FS-UAE.

**Comment by dirkwhoffmann** [comment]
Confirmed. When run with debug options `XFILES` and `BLT_GUARD`, the following output is generated:

```
[1556] (294,213) 0110C4  0 BC--DA 4004 37A2 CPU: XFILES: write16 close to PC 110ce
[1556] (294,215) 0110C4  0 BC--DA 4004 37A2 CPU: XFILES: write16 close to PC 110ce
[2644] (295, 23) 0110C4  0 BC--DA 4004 37A2 CPU: XFILES: write16 close to PC 110ce
[2644] (295, 25) 0110C4  0 BC--DA 4004 37A2 CPU: XFILES: write16 close to PC 110ce
[2650] (191,180) 011204  0 -CBS-- 601C 0000 CPU: XFILES: TAS instruction
[2650] (192, 68) 011204  0 -CBS-- 601C 0000 CPU: XFILES: TAS instruction
...
[2655] ( 28,143) 011204  0 -CBS-A 601C 07A0 CPU: XFILES: TAS instruction
[2655] (232,195) FC081A  0 -CB--A 601C 07A0 CPU: XFILES: Address error exception d0f5 80087 d0f7 2000 9594
```

The problem occurs around frame 2644. According to the debug log, it might be an issue related to self-modifying code. The TAS instruction message can be ignored (it's likely that the CPU has already gone nuts at this point). The Guru comes up, because the CPU eventually causes an address error exception.

**Comment by dirkwhoffmann** [comment]
For reference, the address error is generated here: 

<img width="1036" alt="Bildschirmfoto 2021-01-17 um 14 27 55" src="https://user-images.githubusercontent.com/12561945/104844396-4d677600-58d0-11eb-9784-aa017c7ab790.png">

**Comment by dirkwhoffmann** [comment]
It's a drive issue. It works when a turbo drive is selected: 

<img width="988" alt="Bildschirmfoto 2021-01-22 um 13 43 49" src="https://user-images.githubusercontent.com/12561945/105492253-f2f75c80-5cb7-11eb-9e59-2c0f11739de5.png">

**Comment by dirkwhoffmann** [comment]
More findings: It also works with debug option `ALIGN_HEAD = 1`. This indicates that it has something to with cylinder switching.

I did further experiments in function `Drive::step()`

When 

    head.offset = ALIGN_HEAD ? 0 : head.offset;

is replaced by 

    head.offset = ALIGN_HEAD ? 0 : ((head.offset + 117) % last);

it also works. In plain English, this means the following: The current implementation of vAmiga switches cylinders in no time. When a time delay of 117 bytes is emulated, the game boots just fine. Bottom line: vAmiga needs to emulate the mechanical delay when the head moves from one cylinder to another.

Side note: The experimental fix described here also fixes #411.

Now, the big question is: How long does a cylinder switch take and what happens to the FIFO buffer while the head switches. 🤔

**Comment by dirkwhoffmann** [comment]
By reviewing my own code, I noticed that I have already implemented a head step delay a couple of month ago (I completely forgot about that which means that either I am getting too old, or my code base is getting too large, or both 🙄):

```
u8
Drive::readByte() const
{
    // Case 1: No disk is inserted
    if (!disk) {
        return 0xFF;
    }

    // Case 2: A step operation is in progress
    if (config.mechanicalDelays && (agnus.clock - stepCycle) < config.stepDelay) {
        return (u8)rand(); // 0xFF;
    }
    
    // Case 3: Normal operation
    return disk->readByte(head.cylinder, head.side, head.offset);
}
```

It's just that I used a too small delay in my old implementation:

    config.stepDelay = USEC(2000);

Changing this line to

    config.stepDelay = USEC(8000);

fixes this issue as well as the Terminator 2 issue (#411).

Side note: `USEC(7500)` also works, `USEC(7000)` does not.

**Comment by dirkwhoffmann** [comment]
Fixed in v0.9.17.

**Comment by bnice** [comment]
Hi, I have tried both Midnight Resistance  and Prince of Persia on v0.9.17 with mechanical delays enabled and disabled  can't get either to work :-( I deleted my the preferences and still no luck. Is there any other settings I need to make to allow them to work?

**Comment by dirkwhoffmann** [comment]
It does work with my Xcode build, but it doesn't with the executable from the release page. Maybe, I've uploaded the wrong binary. I'll check...

**Comment by dirkwhoffmann** [comment]
Can you try again with v0.9.17.1? I think I screwed v0.9.17 up 😬.

**Comment by bnice** [comment]
All good both work now. You can also add Pro Tennis Tour 2 to list of games fixed with this release that was the only game I use to have to turn DSync on to work 🙂.

**Comment by dirkwhoffmann** [comment]
> All good both work now

Great! I'll delete the (broken) v0.9.17 executable then.
